File: 0155-min-stack/0155-min-stack.py
class Node:
  def __init__(self, val: int):
    self.val, self.min = val, None

class MinStack:
  def __init__(self):
    self.stack = []

  def push(self, val: int) -> None:
    newNode = Node(val)
    if not self.stack or self.getMin() > val:
      newNode.min = val
    else:
      newNode.min = self.getMin()
    self.stack.append(newNode)

  def pop(self) -> None:
    self.stack.pop()

  def top(self) -> int:
    return self.stack[-1].val

  def getMin(self) -> int:
    return self.stack[-1].min


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: 0155-min-stack/0155-min-stack.ts
class Node {
  val: number
  min: number
  
  constructor(val: number) {
    this.val = val
  }
}

class MinStack {
  stack: Node[] = []

  constructor() {
    
  }

  push(val: number): void {
    const newNode: Node = new Node(val)
    if (!this.stack.length || this.getMin() > val) {
      newNode.min = val
    } else {
      newNode.min = this.getMin()
    }
    this.stack.push(newNode)
  }

  pop(): void {
    this.stack.pop()
  }

  top(): number {
    return this.stack[this.stack.length - 1].val
  }

  getMin(): number {
    return this.stack[this.stack.length - 1].min
  }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(val)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */